c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+/// differ in shape (for example, if `hir_ty` is `_` or a type alias), then the result is `None`.
+/// If they have the same shape, the result is a list of `hir::Ty`s of arguments, where "argument"
+/// is defined the same way as for the `args` field of `LabeledTy`.
+fn deconstruct_hir_ty<'a, 'tcx>(
+    ty: Ty<'tcx>,
+    hir_ty: &'a hir::Ty<'tcx>,
+) -> Option<Vec<&'a hir::Ty<'tcx>>> {
+    /// Match a specific primitive type against a `hir::Ty`.  Returns `Some([])` on success (since
+    /// primitive types have no arguments) or `None` on failure.
+    fn do_prim<'a, 'tcx>(
+        expect_prim: hir::PrimTy,
+        hir_ty: &hir::Ty<'tcx>,
+    ) -> Option<Vec<&'a hir::Ty<'tcx>>> {
+        let path = match hir_ty.kind {
+            hir::TyKind::Path(hir::QPath::Resolved(_, path)) => path,
+            _ => return None,
+        };
+        let prim = match path.res {
+            Res::PrimTy(prim) => prim,
+            _ => return None,
+        };
+        if prim != expect_prim {
+            return None;
+        }
+        Some(Vec::new())
+    }
+
+    match (ty.kind(), &hir_ty.kind) {
+        // Types with no arguments
+        (&TyKind::Bool, _) => do_prim(hir::PrimTy::Bool, hir_ty),
+        (&TyKind::Char, _) => do_prim(hir::PrimTy::Char, hir_ty),
+        (&TyKind::Int(ity), _) => do_prim(hir::PrimTy::Int(convert_int_ty(ity)), hir_ty),
+        (&TyKind::Uint(uty), _) => do_prim(hir::PrimTy::Uint(convert_uint_ty(uty)), hir_ty),
+        (&TyKind::Float(fty), _) => do_prim(hir::PrimTy::Float(convert_float_ty(fty)), hir_ty),
+        (&TyKind::Str, _) => do_prim(hir::PrimTy::Str, hir_ty),
+
+        // Types with arguments
+        (&TyKind::Array(_, _), &hir::TyKind::Array(arg_ty, _)) => Some(vec![arg_ty]),
+        (&TyKind::Slice(_), &hir::TyKind::Slice(arg_ty)) => Some(vec![arg_ty]),
+        (&TyKind::RawPtr(tm), &hir::TyKind::Ptr(ref hir_mt)) => {
+            if hir_mt.mutbl == convert_mutability(tm.mutbl) {
+                Some(vec![hir_mt.ty])
+            } else {
+                None
+            }
+        }
+        (&TyKind::Ref(_, _, mutbl), &hir::TyKind::Rptr(_, ref hir_mt)) => {
+            if hir_mt.mutbl == convert_mutability(mutbl) {
+                Some(vec![hir_mt.ty])
+            } else {
+                None
+            }
+        }
+        (&TyKind::Tuple(tys), &hir::TyKind::Tup(hir_tys)) => {
+            if tys.len() == hir_tys.len() {
+                Some(hir_tys.iter().collect())
+            } else {
+                None
+            }
+        }
+
+        _ => None,
+    }
+}
+
+fn convert_int_ty(x: ty::IntTy) -> ast::IntTy {
+    match x {
+        ty::IntTy::Isize => ast::IntTy::Isize,
+        ty::IntTy::I8 => ast::IntTy::I8,
+        ty::IntTy::I16 => ast::IntTy::I16,
+        ty::IntTy::I32 => ast::IntTy::I32,
+        ty::IntTy::I64 => ast::IntTy::I64,
+        ty::IntTy::I128 => ast::IntTy::I128,
+    }
+}
```

Comment:
I'm surprised things like this aren't already in a `From` `impl`.  We could coalesce them into a locally-defined (for the orphan rule) `trait Convert<T>` or something, which helps with future discoverability.